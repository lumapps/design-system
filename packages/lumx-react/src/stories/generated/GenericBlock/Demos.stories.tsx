/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/generic-block/GenericBlock Demos' };

export { App as Default } from './default';
