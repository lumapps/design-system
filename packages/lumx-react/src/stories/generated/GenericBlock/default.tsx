../../../../../site-demo/content/product/components/generic-block/react/default.tsx