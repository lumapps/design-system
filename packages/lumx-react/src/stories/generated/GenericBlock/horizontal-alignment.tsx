../../../../../site-demo/content/product/components/generic-block/react/horizontal-alignment.tsx