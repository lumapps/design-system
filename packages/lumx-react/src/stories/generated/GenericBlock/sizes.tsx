../../../../../site-demo/content/product/components/generic-block/react/sizes.tsx