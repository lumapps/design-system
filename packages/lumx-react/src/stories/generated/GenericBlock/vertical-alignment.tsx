../../../../../site-demo/content/product/components/generic-block/react/vertical-alignment.tsx