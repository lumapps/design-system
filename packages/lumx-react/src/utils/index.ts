/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
