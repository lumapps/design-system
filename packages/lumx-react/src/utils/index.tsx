export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
