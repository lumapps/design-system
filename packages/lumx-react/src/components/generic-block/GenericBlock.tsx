import React, { Children, forwardRef, ReactElement, ReactNode } from 'react';
import classNames from 'classnames';
import isEmpty from 'lodash/isEmpty';
import noop from 'lodash/noop';
import { Comp, isComponentType } from '@lumx/react/utils/type';
import { getRootClassName } from '@lumx/react/utils/className';
import { partitionMulti } from '@lumx/react/utils/partitionMulti';
import { Orientation, Size, FlexBox, FlexBoxProps } from '@lumx/react';
import { GenericBlockGapSize } from '@lumx/react/components/generic-block/constants';

export interface GenericBlockProps extends FlexBoxProps {
    /**
     * Component to use as visual element.
     */
    figure?: ReactNode;
    /**
     * Actions to set after the main content.
     */
    actions?: ReactNode;
    /**
     * Main content to display or sections components
     * ({@see GenericBlock.Figure}, {@see GenericBlock.Content} & {@see GenericBlock.Actions})
     */
    children: ReactNode;
    /**
     * Orientation of the 3 sections
     */
    orientation?: FlexBoxProps['orientation'];
    /**
     * Horizontal alignment.
     */
    hAlign?: FlexBoxProps['hAlign'];
    /**
     * Vertical alignment.
     */
    vAlign?: FlexBoxProps['vAlign'];
    /**
     * The props to forward to the content.
     * By default, the content will have the same alignment as wrapper.
     */
    contentProps?: Omit<FlexBoxProps, 'children'>;
    /**
     * props to forward to the actions element.
     */
    actionsProps?: Omit<FlexBoxProps, 'children'>;
    /**
     * props to forward to the figure element.
     */
    figureProps?: Omit<FlexBoxProps, 'children'>;
    /**
     * Gap space between sections.
     */
    gap?: GenericBlockGapSize;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'GenericBlock';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<GenericBlockProps> = {
    gap: Size.big,
    orientation: Orientation.horizontal,
};

type BaseGenericBlock = Comp<GenericBlockProps, HTMLDivElement>;

interface GenericBlockSectionProps extends FlexBoxProps {
    /**
     * Gap space between items.
     */
    gap?: GenericBlockGapSize;
}

interface GenericBlock extends BaseGenericBlock {
    /**
     * Use `GenericBlock.Figure` component as children of the `GenericBlock` component as an alternative way to inject
     * the "figure" section of the block (instead of using `figure` and `figureProps` props).
     */
    Figure: Comp<GenericBlockSectionProps>;
    /**
     * Use `GenericBlock.Content` component as children of the `GenericBlock` component as an alternative way to inject
     * the "content" section of the block (instead of using `content` and `contentProps` props).
     */
    Content: Comp<GenericBlockSectionProps>;
    /**
     * Use `GenericBlock.Actions` component as children of the `GenericBlock` component as an alternative way to inject
     * the "actions" section of the block (instead of using `actions` and `actionsProps` props).
     */
    Actions: Comp<GenericBlockSectionProps>;
}

const Figure = noop.bind({}) as Comp<FlexBoxProps>;
const isFigure = isComponentType(Figure);

const Content = noop.bind({}) as Comp<FlexBoxProps>;
const isContent = isComponentType(Content);

const Actions = noop.bind({}) as Comp<FlexBoxProps>;
const isActions = isComponentType(Actions);

/**
 * The GenericBlock is a layout component made of 3 sections that can be
 * displayed either horizontally of vertically with the same gap between each section.
 *
 * The sections are:
 * - `Figure` => A visual element to display before the main content.
 * - `Content` => The main content displayed
 * - `Actions` => One or more actions to set after the element.
 *
 * @see https://www.figma.com/file/lzzrQmsfaXRaOyRfoEogPZ/DS%3A-playground?node-id=1%3A4076
 */
const BaseGenericBlock: BaseGenericBlock = forwardRef((props, ref) => {
    const {
        className,
        figure,
        figureProps,
        children,
        actions,
        actionsProps,
        gap,
        orientation,
        contentProps,
        ...forwardedProps
    } = props;

    const sections = React.useMemo(() => {
        // Split children by section type
        const [[figureChild], [contentChild], [actionsChild], ...otherChildren] = partitionMulti(
            Children.toArray(children),
            [isFigure, isContent, isActions],
        );
        return {
            figureChild,
            figureChildProps: (figureChild as ReactElement)?.props,
            contentChild,
            contentChildProps: (contentChild as ReactElement)?.props,
            actionsChild,
            actionsChildProps: (actionsChild as ReactElement)?.props,
            otherChildren: otherChildren.filter((child) => !isEmpty(child)),
        };
    }, [children]);

    return (
        <FlexBox
            ref={ref}
            className={classNames(className, CLASSNAME)}
            gap={gap}
            orientation={orientation}
            {...forwardedProps}
        >
            {(figure || sections.figureChildProps?.children) && (
                <FlexBox
                    ref={(sections.figureChild as any)?.ref}
                    vAlign={forwardedProps.vAlign}
                    hAlign={forwardedProps.hAlign}
                    {...figureProps}
                    {...sections.figureChildProps}
                    className={classNames(
                        figureProps?.className,
                        sections.figureChildProps?.className,
                        `${CLASSNAME}__figure`,
                    )}
                >
                    {figure}
                    {sections.figureChildProps?.children}
                </FlexBox>
            )}

            {(sections.contentChildProps?.children || sections.otherChildren.length > 0) && (
                <FlexBox
                    ref={(sections.contentChild as any)?.ref}
                    orientation={Orientation.vertical}
                    fillSpace
                    vAlign={forwardedProps.vAlign}
                    hAlign={forwardedProps.hAlign}
                    {...contentProps}
                    {...sections.contentChildProps}
                    className={classNames(
                        contentProps?.className,
                        sections.contentChildProps?.className,
                        `${CLASSNAME}__content`,
                    )}
                >
                    {sections.contentChildProps?.children}
                    {sections.otherChildren}
                </FlexBox>
            )}

            {(actions || sections.actionsChildProps?.children) && (
                <FlexBox
                    ref={(sections.actionsChild as any)?.ref}
                    vAlign={forwardedProps.vAlign}
                    hAlign={forwardedProps.hAlign}
                    {...actionsProps}
                    {...sections.actionsChildProps}
                    className={classNames(
                        actionsProps?.className,
                        sections.actionsChildProps?.className,
                        `${CLASSNAME}__actions`,
                    )}
                >
                    {actions}
                    {sections.actionsChildProps?.children}
                </FlexBox>
            )}
        </FlexBox>
    );
});
BaseGenericBlock.displayName = COMPONENT_NAME;
BaseGenericBlock.className = CLASSNAME;
BaseGenericBlock.defaultProps = DEFAULT_PROPS;

export const GenericBlock: GenericBlock = Object.assign(BaseGenericBlock, { Figure, Content, Actions });
