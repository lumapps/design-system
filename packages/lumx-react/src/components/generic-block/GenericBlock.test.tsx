import React, { createRef } from 'react';
import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { commonTestsSuite } from '@lumx/react/testing/utils';

import { GenericBlock, GenericBlockProps } from './GenericBlock';

const CLASSNAME = GenericBlock.className as string;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (props: Partial<GenericBlockProps> = {}, shallowRendering = true) => {
    const renderer: any = shallowRendering ? shallow : mount;
    const wrapper: any = renderer(<GenericBlock {...(props as any)} />);
    return { props, wrapper };
};

describe(`<${GenericBlock.displayName}>`, () => {
    // 1. Test render via snapshot.
    describe('Snapshots and structure', () => {
        it('should render default', () => {
            const wrapper = shallow(<GenericBlock>Content</GenericBlock>);

            expect(wrapper.prop('gap')).toBe('regular');
            expect(wrapper.prop('hAlign')).toBe('center');
            expect(wrapper.prop('orientation')).toBe('vertical');
            expect(wrapper.prop('vAlign')).toBe('center');

            const figure = wrapper.find('.lumx-generic-block__figure');
            expect(figure).toBeDefined();

            const content = wrapper.find('.lumx-generic-block__content');
            expect(content.prop('fillSpace')).toBe(true);
            expect(content.prop('vAlign')).toBe('center');
            expect(content.prop('orientation')).toBe('vertical');

            const actions = wrapper.find('.lumx-generic-block__actions');
            expect(actions.prop('vAlign')).toBe('center');
        });

        it('should render horizontal', () => {
            const wrapper = shallow(<GenericBlock orientation="horizontal">Content</GenericBlock>);

            expect(wrapper.prop('gap')).toBe('regular');
            expect(wrapper.prop('hAlign')).toBe('center');
            expect(wrapper.prop('orientation')).toBe('horizontal');
            expect(wrapper.prop('vAlign')).toBe('center');

            const figure = wrapper.find('.lumx-generic-block__figure');
            expect(figure).toBeDefined();

            const content = wrapper.find('.lumx-generic-block__content');
            expect(content.prop('fillSpace')).toBe(true);
            expect(content.prop('vAlign')).toBe('left');
            expect(content.prop('orientation')).toBe('vertical');

            const actions = wrapper.find('.lumx-generic-block__actions');
            expect(actions.prop('vAlign')).toBe('right');
        });

        it('should combine figure props', () => {
            const wrapper = shallow(
                <GenericBlock figure="Figure 1" figureProps={{ className: 'figure1', vAlign: 'left' }}>
                    <GenericBlock.Figure className="figure2" fillSpace>
                        Figure 2
                    </GenericBlock.Figure>
                </GenericBlock>,
            );

            const figure = wrapper.find('.lumx-generic-block__figure');
            expect(figure).toHaveClassName('figure1');
            expect(figure).toHaveClassName('figure2');
            expect(figure.prop('fillSpace')).toBe(true);
            expect(figure.prop('vAlign')).toBe('left');
            expect(figure).toHaveText('Figure 1Figure 2');
        });

        it('should combine content props', () => {
            const wrapper = shallow(
                <GenericBlock contentProps={{ className: 'content1', vAlign: 'left' }}>
                    Content 1
                    <GenericBlock.Content className="content2" fillSpace>
                        Content 2
                    </GenericBlock.Content>
                </GenericBlock>,
            );

            const content = wrapper.find('.lumx-generic-block__content');
            expect(content).toHaveClassName('content1');
            expect(content).toHaveClassName('content2');
            expect(content.prop('fillSpace')).toBe(true);
            expect(content.prop('vAlign')).toBe('left');
            expect(content).toHaveText('Content 2Content 1');
        });

        it('should combine actions props', () => {
            const wrapper = shallow(
                <GenericBlock actions="Actions 1" actionsProps={{ className: 'actions1', vAlign: 'left' }}>
                    <GenericBlock.Actions className="actions2" fillSpace>
                        Actions 2
                    </GenericBlock.Actions>
                </GenericBlock>,
            );

            const actions = wrapper.find('.lumx-generic-block__actions');
            expect(actions).toHaveClassName('actions1');
            expect(actions).toHaveClassName('actions2');
            expect(actions.prop('fillSpace')).toBe(true);
            expect(actions.prop('vAlign')).toBe('left');
            expect(actions).toHaveText('Actions 1Actions 2');
        });

        it('should forward refs', () => {
            const rootRef = createRef<HTMLDivElement>();
            const figureRef = createRef<HTMLDivElement>();
            const contentRef = createRef<HTMLDivElement>();
            const actionsRef = createRef<HTMLDivElement>();
            shallow(
                <GenericBlock ref={rootRef}>
                    <GenericBlock.Figure ref={figureRef} />
                    <GenericBlock.Content ref={contentRef} />
                    <GenericBlock.Actions ref={actionsRef} />
                </GenericBlock>,
            );
            expect(rootRef.current).toBeDefined();
            expect(figureRef.current).toBeDefined();
            expect(contentRef.current).toBeDefined();
            expect(actionsRef.current).toBeDefined();
        });
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
