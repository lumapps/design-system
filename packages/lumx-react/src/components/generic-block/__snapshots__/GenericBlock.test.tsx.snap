// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericBlock> Snapshots and structure should render story 'Default' 1`] = `
Array [
  <FlexBox
    className="lumx-generic-block"
    gap="regular"
    hAlign="center"
    orientation="vertical"
    vAlign="center"
  >
    <FlexBox
      className="lumx-generic-block__figure"
    >
      <Avatar
        alt=""
        image="/demo-assets/persona.png"
        size="m"
        theme="light"
      />
    </FlexBox>
    <FlexBox
      className="lumx-generic-block__content"
      fillSpace={true}
      orientation="vertical"
      vAlign="center"
    >
      <h2>
        Content title
      </h2>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec rhoncus libero aliquet pharetra luctus. Fusce nisl turpis, posuere ac tellus at, euismod vulputate libero...
      </p>
    </FlexBox>
    <FlexBox
      className="lumx-generic-block__actions"
      vAlign="center"
    >
      <Button
        emphasis="high"
        size="m"
        theme="light"
      >
        Actions
      </Button>
    </FlexBox>
  </FlexBox>,
  <FlexBox
    className="lumx-generic-block"
    gap="regular"
    hAlign="center"
    orientation="horizontal"
    vAlign="center"
  >
    <FlexBox
      className="lumx-generic-block__figure"
    >
      <Avatar
        alt=""
        image="/demo-assets/persona.png"
        size="m"
        theme="light"
      />
    </FlexBox>
    <FlexBox
      className="lumx-generic-block__content"
      fillSpace={true}
      orientation="vertical"
      vAlign="left"
    >
      <h2>
        Content title
      </h2>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec rhoncus libero aliquet pharetra luctus. Fusce nisl turpis, posuere ac tellus at, euismod vulputate libero...
      </p>
    </FlexBox>
    <FlexBox
      className="lumx-generic-block__actions"
      vAlign="right"
    >
      <Button
        emphasis="high"
        size="m"
        theme="light"
      >
        Actions
      </Button>
    </FlexBox>
  </FlexBox>,
]
`;
